memchunk_tpl.h

```diff
@@ -0,0 +1,166 @@
+        }
+        if (len & 4) {
+            memcpy(out, from, 4);
+            out += 4;
+            from += 4;
+        }
+        if (len & 2) {
+            memcpy(out, from, 2);
+            out += 2;
+            from += 2;
+        }
+        if (len & 1) {
+            *out++ = *from++;
+        }
+        return out;
+    }
+    return CHUNKCOPY(out, from, len);
+}
+
+/* Perform short copies until distance can be rewritten as being at least
+   CHUNKMEMSET_SIZE.
+
+   This assumes that it's OK to overwrite at least the first
+   2*CHUNKMEMSET_SIZE bytes of output even if the copy is shorter than this.
+   This assumption holds because inflate_fast() starts every iteration with at
+   least 258 bytes of output space available (258 being the maximum length
+   output from a single token; see inflate_fast()'s assumptions below). */
+uint8_t* CHUNKUNROLL(uint8_t *out, unsigned *dist, unsigned *len) {
+    unsigned char const *from = out - *dist;
+    memchunk_t chunk;
+    while (*dist < *len && *dist < CHUNKMEMSET_SIZE) {
+        loadchunk(from, &chunk);
+        storechunk(out, &chunk);
+        out += *dist;
+        *len -= *dist;
+        *dist += *dist;
+    }
+    return out;
+}
+
+/* Copy DIST bytes from OUT - DIST into OUT + DIST * k, for 0 <= k < LEN/DIST.
+   Return OUT + LEN. */
+uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
+    /* Debug performance related issues when len < sizeof(uint64_t):
+       Assert(len >= sizeof(uint64_t), "chunkmemset should be called on larger chunks"); */
+    Assert(dist > 0, "cannot have a distance 0");
+
+    unsigned char *from = out - dist;
+    memchunk_t chunk;
+    unsigned sz = sizeof(chunk);
+    if (len < sz) {
+        do {
+            *out++ = *from++;
+            --len;
+        } while (len != 0);
+        return out;
+    }
+
+#ifdef HAVE_CHUNKMEMSET_1
+    if (dist == 1) {
+        chunkmemset_1(from, &chunk);
+    } else
+#endif
+#ifdef HAVE_CHUNKMEMSET_2
+    if (dist == 2) {
+        chunkmemset_2(from, &chunk);
+    } else
+#endif
+#ifdef HAVE_CHUNKMEMSET_3
+    if (dist == 3) {
+        return chunkmemset_3(out, from, dist, len);
+    } else
+#endif
+#ifdef HAVE_CHUNKMEMSET_4
+    if (dist == 4) {
+        chunkmemset_4(from, &chunk);
+    } else
+#endif
+#ifdef HAVE_CHUNKMEMSET_6
+    if (dist == 6) {
+        return chunkmemset_6(out, from, dist, len);
+    } else
+#endif
+#ifdef HAVE_CHUNKMEMSET_8
+    if (dist == 8) {
+        chunkmemset_8(from, &chunk);
+    } else
+#endif
+    if (dist == sz) {
+        loadchunk(from, &chunk);
+    } else {
+        out = CHUNKUNROLL(out, &dist, &len);
+        return CHUNKCOPY(out, out - dist, len);
+    }
+
+    unsigned rem = len % sz;
+    len -= rem;
+    while (len) {
+        storechunk(out, &chunk);
+        out += sz;
+        len -= sz;
+    }
+
+    /* Last, deal with the case when LEN is not a multiple of SZ. */
+    if (rem)
+        memcpy(out, from, rem);
+    out += rem;
+
+    return out;
+}
```

Comment:
End of the function might be faster if added as inlined function so compiler does not need to emit jump from each if condition above... Code size in compiled form might increase a bit though. This also reduced size and complexity of this function in source code form. Inlined function should return "out + len" so it can be called and used with "return" directly.

Reply:
I think that might be better for a future optimization.